src/main/java/net/fabricmc/fabric/api/fluid/FluidMeter.java

```diff
@@ -0,0 +1,108 @@
+package net.fabricmc.fabric.api.fluid;
+
+import net.minecraft.nbt.LongArrayTag;
+import net.minecraft.util.PacketByteBuf;
+
+/**
+ * For use within a tank, machine, item or other game entity to track
+ * the stored amount of a fluid or other volumetric substance measured
+ * and transfered using {@link FluidUnit}.<p>
+ * 
+ * By design, it serves only to handle the accounting of contents.
+ * Assumes all units tracked are for the <em>same</em> fluid. (Tracking
+ * multiple fluids in the same device will require more than one {@link FluidMeter}
+ * but otherwise does not know anything about the tracked fluid or its behavior.<p>
+ * 
+ * The enclosing device may constrain the units allowed for transfer in or out
+ * depending on the fluid itself and expected use cases.<p>
+ * 
+ * The enclosing device is also responsible for transfer semantics such as capacity limits, 
+ * simulated operations, etc. (TODO: those are likely to be a separate part of this API.)
+ * 
+ * TODO: static factory for creating/loading instances.
+ */
+public interface FluidMeter {
+    /**
+     * Adds fluid to the internal tally, returning the amount added. The meter can track at 
+     * least {@link Long#MAX_VALUE} buckets of fluid.  Any attempt to add more than the implemented 
+     * limit will result in a partial addition.
+     */
+    long add(long amount, FluidUnit unit);
+    
+    /**
+     * Removes fluid from the internal tally, up to the given amount. If the amount given exceeds
+     * the current tally, only the present amount will be removed and that amount returned as the result.
+     * (Negative amounts can never occur.)<p>
+     */
+    long remove(long amount, FluidUnit unit);
+    
+    /**
+     * Returns the current fluid amount in the chosen unit. The integer portion of the return value
+     * will be exact unless the total is larger than double precision can support. 
+     */
+    double total(FluidUnit unit);
+    
+    /**
+     * Returns current fluid amount in buckets, excluding any fractional amount. Always exact. 
+     */
+    long buckets();
```

Comment:
I don't think longs are really necessary for the amount of buckets. I can't really think of any time you'd ever have 2^31-1 buckets of any fluid in one place, much less more than that.

Reply:
*Should* someone make a tank that can store more than 32,768 full world chunks of fluid?

Maybe not, but I’m going to serialize to a long array anyway so may as well. It makes any potential complaining about lack of scale positively ridiculous.

Reply:
I'm just worried that on a busy server all those extra unused bytes for *every fluid meter* are gonna add up.